README
NAME
    Image::Select - Selecting image from images directory.

SYNOPSIS
     use Image::Select;
     my $obj = Image::Select->new(%parameters);
     my $type = $obj->create($output_path);
     my ($width, $height) = $obj->sizes($new_width, $new_height);
     my $type = $obj->type($new_type);

METHODS
    "new(%parameters)"
             Constructor.

            *       "debug"

                     Debug mode.
                     Default value is 0.

            *       "height"

                     Height of image.
                     Default value is 1920.

            *       "loop"

                     Returns images in loop.
                     Default value is 0.

            *       "path_to_images"

                     Path to images.
                     It is required.
                     Default value is undef.

            *       "type"

                     Image type.
                     List of supported types: bmp, gif, jpeg, png, pnm, raw, sgi, tga, tiff
                     Default value is undef.

            *       "width"

                     Width of image.
                     Default value is 1080.

    "create($path)"
             Create image.
             Returns scalar value of supported file type.

    "sizes([$width, $height])"
             Set/Get image sizes.
             Returns actual width and height.

    "type([$type])"
             Set/Get image type.
             Returns actual type of image.

ERRORS
     new():
             No images.
             Parameter 'path_to_images' is required.
             Image type '%s' doesn't supported.
             Class::Utils:
                     Unknown parameter '%s'.

     create():
             Cannot read file '%s'.
                     Error, %s
             Cannot resize image from file '%s'.
                     Error, %s
             Cannot write file to '$path'.
                     Error, %s
             No file '%s'.
             Image type '%s' doesn't supported.

EXAMPLE1
     use strict;
     use warnings;

     use File::Spec::Functions qw(catfile);
     use File::Temp qw(tempfile tempdir);
     use Image::Random;
     use Image::Select;

     # Temporary directory to random images.
     my $tempdir = tempdir(CLEANUP => 1);

     # Create temporary images.
     my $rand = Image::Random->new;
     for my $i (1 .. 5) {
             $rand->create(catfile($tempdir, $i.'.png'));
     }

     # Object.
     my $obj = Image::Select->new(
             'path_to_images' => $tempdir,
     );

     # Temporary file.
     my (undef, $temp) = tempfile();

     # Create image.
     my $type = $obj->create($temp);

     # Print out type.
     print $type."\n";

     # Unlink file.
     unlink $temp;

     # Output:
     # bmp

EXAMPLE2
     use strict;
     use warnings;

     use File::Spec::Functions qw(catfile);
     use File::Temp qw(tempfile tempdir);
     use Image::Random;
     use Image::Select;

     # Temporary directory for random images.
     my $tempdir = tempdir(CLEANUP => 1);

     # Create temporary images.
     my $rand = Image::Random->new;
     for my $i (1 .. 5) {
             $rand->create(catfile($tempdir, $i.'.png'));
     }

     # Object.
     my $obj = Image::Select->new(
             'loop' => 0,
             'path_to_images' => $tempdir,
     );

     # Temporary file.
     my (undef, $temp) = tempfile();

     # Create image.
     while (my $type = $obj->create($temp)) {

             # Print out type.
             print $type."\n";
     }

     # Unlink file.
     unlink $temp;

     # Output:
     # bmp
     # bmp
     # bmp
     # bmp
     # bmp

DEPENDENCIES
    Class::Utils, Error::Pure, File::Basename, File::Find::Rule,
    File::Find::Rule::MMagic, Imager, List::MoreUtils.

SEE ALSO
    Image::Random
        Perl class for creating random image.

    Image::Select::Array
        Selecting image from list with checking.

    Image::Select::Date
        Selecting image from images directory by date.

REPOSITORY
    <https://github.com/michal-josef-spacek/Image-Select>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
     © 2014-2020 Michal Josef Špaček
     BSD 2-Clause License

VERSION
    0.06



Changes
0.06

0.05 2020-01-18T18:42:18+01:00
 - Add Image::Select::Date to SEE ALSO section in doc.
 - Fix Makefile.PL fails with no '.' in @INC.
 - Update .travis.yml to reality.
 - Update Module::Install to 1.19 version.
 - Update author github username.
 - Update author name.
 - Update copyright years.

0.04 2015-11-19T11:59:09+01:00
 - Add Image::Select::Array module.
 - Fix and improve ABSTRACT.

0.03 2015-09-30T16:28:48+02:00
 - Add REPOSITORY section to doc.
 - Fix test requires.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SEE ALSO section in doc.
 - Move author tests to xt/ directory.
 - Update Module::Install to 1.16 version.
 - Update copyright years.

0.02 2014-10-25T09:20:52+02:00
 - Add error check for 'path_to_images' parameter.
 - Add tests.
 - Fix dependencies.
 - Rewrite word "Suffix" to "Image type" in error message.
 - Sort modules in EXAMPLE sections in doc.

0.01 2014-10-24T20:24:38+02:00
 - First version.


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Selecting image from images directory.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.05;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'Image-Select';
readme_from 'Select.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0.03;
requires 'Error::Pure' => 0.11;
requires 'File::Find::Rule' => 0;
requires 'File::Find::Rule::MMagic' => 0;
requires 'File::Basename' => 0;
requires 'Imager' => 0;
requires 'List::MoreUtils' => 0;
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://rt.cpan.org/Public/Dist/Display.html'.
	'?Name=Image-Select';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Image-Select';
resources 'repository' => 'git://github.com/michal-josef-spacek/Image-Select';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0;
test_requires 'File::Object' => 0.05;
test_requires 'Imager::Color' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.06';

# Run.
WriteAll();


Select/Array.pm
package Image::Select::Array;

use base qw(Image::Select);
use strict;
use warnings;

use Class::Utils qw(set_params);
use Error::Pure qw(err);

our $VERSION = 0.06;

sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Debug.
	$self->{'debug'} = 0;

	# Loop.
	$self->{'loop'} = 0;

	# Image list.
	$self->{'image_list'} = [];

	# Image type.
	$self->{'type'} = 'bmp';

	# Sizes.
	$self->{'height'} = 1080;
	$self->{'width'} = 1920;

	# Process params.
	set_params($self, @params);

	# Check type.
	if (defined $self->{'type'}) {
		$self->_check_type($self->{'type'});
	}

	# Check images array.
	if (! defined $self->{'image_list'}
		|| ref $self->{'image_list'} ne 'ARRAY') {

		err "Parameter 'image_list' must be reference to array ".
			'with images.';
	}

	# Check images.
	if (! @{$self->{'image_list'}}) {
		err 'No images.';
	}

	# Check images path.
	foreach my $image (@{$self->{'image_list'}}) {
		if (! -r $image) {
			err "Image '$image' doesn't readable.";
		}
	}

	# Images to select.
	$self->{'_images_to_select'} = $self->{'image_list'};
	$self->{'_images_index'} = 0;

	# Object.
	return $self;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Image::Select::Array - Selecting image from list with checking.

=head1 SYNOPSIS

 use Image::Select::Array;
 my $obj = Image::Select::Array->new(%parameters);
 my $type = $obj->create($output_path);
 my ($width, $height) = $obj->sizes($new_width, $new_height);
 my $type = $obj->type($new_type);

=head1 METHODS

=over 8

=item C<new(%parameters)>

 Constructor.

=over 8

=item * C<debug>

 Debug mode.
 Default value is 0.

=item * C<height>

 Height of image.
 Default value is 1920.

=item * C<loop>

 Returns images in loop.
 Default value is 0.

=item * C<image_list>

 List of images in array reference.
 It is required.
 Default value is [].

=item * C<type>

 Image type.
 List of supported types: bmp, gif, jpeg, png, pnm, raw, sgi, tga, tiff
 Default value is undef.

=item * C<width>

 Width of image.
 Default value is 1080.

=back

=item C<create($path)>

 Create image.
 Returns scalar value of supported file type.

=item C<sizes([$width, $height])>

 Set/Get image sizes.
 Returns actual width and height.

=item C<type([$type])>

 Set/Get image type.
 Returns actual type of image.

=back

=head1 ERRORS

 new():
         No images.
         Image '%s' doesn't readable.
         Image type '%s' doesn't supported.
         Parameter 'image_list' must be reference to array with images.
         Class::Utils:
                 Unknown parameter '%s'.

 create():
         Cannot read file '%s'.
                 Error, %s
         Cannot resize image from file '%s'.
                 Error, %s
         Cannot write file to '$path'.
                 Error, %s
         No file '%s'.
         Image type '%s' doesn't supported.

=head1 EXAMPLE1

 use strict;
 use warnings;

 use File::Spec::Functions qw(catfile);
 use File::Temp qw(tempfile tempdir);
 use Image::Random;
 use Image::Select::Array;

 # Temporary directory to random images.
 my $tempdir = tempdir(CLEANUP => 1);

 # Create temporary images.
 my $rand = Image::Random->new;
 my @images;
 for my $i (1 .. 5) {
         my $image = catfile($tempdir, $i.'.png');
         $rand->create($image);
         push @images, $image;
 }

 # Object.
 my $obj = Image::Select::Array->new(
         'image_list' => \@images,
 );

 # Temporary file.
 my (undef, $temp) = tempfile();

 # Create image.
 my $type = $obj->create($temp);

 # Print out type.
 print $type."\n";

 # Unlink file.
 unlink $temp;

 # Output:
 # bmp

=head1 EXAMPLE2

 use strict;
 use warnings;

 use File::Spec::Functions qw(catfile);
 use File::Temp qw(tempfile tempdir);
 use Image::Random;
 use Image::Select::Array;

 # Temporary directory for random images.
 my $tempdir = tempdir(CLEANUP => 1);

 # Create temporary images.
 my $rand = Image::Random->new;
 my @images;
 for my $i (1 .. 5) {
         my $image = catfile($tempdir, $i.'.png');
         $rand->create(catfile($tempdir, $i.'.png'));
         push @images, $image;
 }

 # Object.
 my $obj = Image::Select::Array->new(
         'image_list' => \@images,
         'loop' => 0,
 );

 # Temporary file.
 my (undef, $temp) = tempfile();

 # Create image.
 while (my $type = $obj->create($temp)) {

         # Print out type.
         print $type."\n";
 }

 # Unlink file.
 unlink $temp;

 # Output:
 # bmp
 # bmp
 # bmp
 # bmp
 # bmp

=head1 DEPENDENCIES

L<Class::Utils>,
L<Error::Pure>,
L<Image::Select>.

=head1 SEE ALSO

=over

=item L<Image::Random>

Perl class for creating random image.

=item L<Image::Select>

Selecting image from images directory.

=item L<Image::Select::Date>

Selecting image from images directory by date.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Image-Select>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

 © 2015-2020 Michal Josef Špaček
 BSD 2-Clause License

=head1 VERSION

0.06

=cut


Select.pm
package Image::Select;

use strict;
use warnings;

use Class::Utils qw(set_params);
use Error::Pure qw(err);
use File::Basename qw(fileparse);
use File::Find::Rule qw(:MMagic);
use Imager;
use List::MoreUtils qw(none);

our $VERSION = 0.06;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Debug.
	$self->{'debug'} = 0;

	# Loop.
	$self->{'loop'} = 0;

	# Path to images.
	$self->{'path_to_images'} = undef;

	# Image type.
	$self->{'type'} = 'bmp';

	# Sizes.
	$self->{'height'} = 1080;
	$self->{'width'} = 1920;

	# Process params.
	set_params($self, @params);

	# Check type.
	if (defined $self->{'type'}) {
		$self->_check_type($self->{'type'});
	}

	# Check path to images.
	if (! defined $self->{'path_to_images'}
		|| ! -d $self->{'path_to_images'}) {

		err "Parameter 'path_to_images' is required.";
	}

	# Load images.
	$self->{'_images_to_select'} = [
		sort File::Find::Rule->file->magic(
			'image/bmp',
			'image/gif',
			'image/jpeg',
			'image/png',
			'image/tiff',
			'image/x-ms-bmp',
			'image/x-portable-pixmap',
			# XXX tga?
			# XXX raw?
			# XXX sgi?
		)->in($self->{'path_to_images'}),
	];
	if (! @{$self->{'_images_to_select'}}) {
		err 'No images.';
	}
	$self->{'_images_index'} = 0;

	# Object.
	return $self;
}

# Create image.
sub create {
	my ($self, $path) = @_;

	# Load next image.
	my $i = Imager->new;
	if ($self->{'_images_index'} > $#{$self->{'_images_to_select'}}) {
		if ($self->{'loop'}) {
			$self->{'_images_index'} = 0;
		} else {
			return;
		}
	}
	my $file = $self->{'_images_to_select'}->[$self->{'_images_index'}];
	if (! -r $file) {
		err "No file '$file'.";
	}
	my $ret = $i->read('file' => $file);
	if (! $ret) {
		err "Cannot read file '$file'.",
			'Error', Imager->errstr;
	}
	$self->{'_images_index'}++;

	# Get type.
	my $suffix;
	if (! defined $self->{'type'}) {

		# Get suffix.
		(my $name, undef, $suffix) = fileparse($path, qr/\.[^.]*/ms);
		$suffix =~ s/^\.//ms;

		# Jpeg.
		if ($suffix eq 'jpg') {
			$suffix = 'jpeg';
		}

		# Check type.
		$self->_check_type($suffix);
	} else {
		$suffix = $self->{'type'};
	}

	# Scale.
	my $new_i = $i->scale(
		'xpixels' => $self->{'width'},
		'ypixels' => $self->{'height'},
	);
	if (! $new_i) {
		err "Cannot resize image from file '$file'.",
			'Error', Imager->errstr;
	}

	# Save.
	if ($self->{'debug'}) {
		print "Path: $path\n";
	}
	$ret = $new_i->write(
		'file' => $path,
		'type' => $suffix,
	);
	if (! $ret) {
		err "Cannot write file to '$path'.",
			'Error', Imager->errstr;
	}

	return $suffix;
}

# Set/Get image sizes.
sub sizes {
	my ($self, $width, $height) = @_;
	if ($width && $height) {
		$self->{'width'} = $width;
		$self->{'height'} = $height;
	}
	return ($self->{'width'}, $self->{'height'});
}

# Set/Get image type.
sub type {
	my ($self, $type) = @_;
	if ($type) {
		$self->_check_type($type);
		$self->{'type'} = $type;
	}
	return $self->{'type'};
}

# Check supported image type.
sub _check_type {
	my ($self, $type) = @_;
	
	# Check type.
	if (none { $type eq $_ } ('bmp', 'gif', 'jpeg', 'png',
		'pnm', 'raw', 'sgi', 'tga', 'tiff')) {

		err "Image type '$type' doesn't supported.";
	}

	return;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Image::Select - Selecting image from images directory.

=head1 SYNOPSIS

 use Image::Select;
 my $obj = Image::Select->new(%parameters);
 my $type = $obj->create($output_path);
 my ($width, $height) = $obj->sizes($new_width, $new_height);
 my $type = $obj->type($new_type);

=head1 METHODS

=over 8

=item C<new(%parameters)>

 Constructor.

=over 8

=item * C<debug>

 Debug mode.
 Default value is 0.

=item * C<height>

 Height of image.
 Default value is 1920.

=item * C<loop>

 Returns images in loop.
 Default value is 0.

=item * C<path_to_images>

 Path to images.
 It is required.
 Default value is undef.

=item * C<type>

 Image type.
 List of supported types: bmp, gif, jpeg, png, pnm, raw, sgi, tga, tiff
 Default value is undef.

=item * C<width>

 Width of image.
 Default value is 1080.

=back

=item C<create($path)>

 Create image.
 Returns scalar value of supported file type.

=item C<sizes([$width, $height])>

 Set/Get image sizes.
 Returns actual width and height.

=item C<type([$type])>

 Set/Get image type.
 Returns actual type of image.

=back

=head1 ERRORS

 new():
         No images.
         Parameter 'path_to_images' is required.
         Image type '%s' doesn't supported.
         Class::Utils:
                 Unknown parameter '%s'.

 create():
         Cannot read file '%s'.
                 Error, %s
         Cannot resize image from file '%s'.
                 Error, %s
         Cannot write file to '$path'.
                 Error, %s
         No file '%s'.
         Image type '%s' doesn't supported.

=head1 EXAMPLE1

 use strict;
 use warnings;

 use File::Spec::Functions qw(catfile);
 use File::Temp qw(tempfile tempdir);
 use Image::Random;
 use Image::Select;

 # Temporary directory to random images.
 my $tempdir = tempdir(CLEANUP => 1);

 # Create temporary images.
 my $rand = Image::Random->new;
 for my $i (1 .. 5) {
         $rand->create(catfile($tempdir, $i.'.png'));
 }

 # Object.
 my $obj = Image::Select->new(
         'path_to_images' => $tempdir,
 );

 # Temporary file.
 my (undef, $temp) = tempfile();

 # Create image.
 my $type = $obj->create($temp);

 # Print out type.
 print $type."\n";

 # Unlink file.
 unlink $temp;

 # Output:
 # bmp

=head1 EXAMPLE2

 use strict;
 use warnings;

 use File::Spec::Functions qw(catfile);
 use File::Temp qw(tempfile tempdir);
 use Image::Random;
 use Image::Select;

 # Temporary directory for random images.
 my $tempdir = tempdir(CLEANUP => 1);

 # Create temporary images.
 my $rand = Image::Random->new;
 for my $i (1 .. 5) {
         $rand->create(catfile($tempdir, $i.'.png'));
 }

 # Object.
 my $obj = Image::Select->new(
         'loop' => 0,
         'path_to_images' => $tempdir,
 );

 # Temporary file.
 my (undef, $temp) = tempfile();

 # Create image.
 while (my $type = $obj->create($temp)) {

         # Print out type.
         print $type."\n";
 }

 # Unlink file.
 unlink $temp;

 # Output:
 # bmp
 # bmp
 # bmp
 # bmp
 # bmp

=head1 DEPENDENCIES

L<Class::Utils>,
L<Error::Pure>,
L<File::Basename>,
L<File::Find::Rule>,
L<File::Find::Rule::MMagic>,
L<Imager>,
L<List::MoreUtils>.

=head1 SEE ALSO

=over

=item L<Image::Random>

Perl class for creating random image.

=item L<Image::Select::Array>

Selecting image from list with checking.

=item L<Image::Select::Date>

Selecting image from images directory by date.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Image-Select>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

 © 2014-2020 Michal Josef Špaček
 BSD 2-Clause License

=head1 VERSION

0.06

=cut
